p2p/wdrr_queue.go

```diff
@@ -0,0 +1,301 @@
+	enqueueCloser *tmsync.Closer
+	dequeueCh     chan Envelope
+	dequeueCloser *tmsync.Closer
+}
+
+func newWDRRQueue(logger log.Logger, pID NodeID, m *Metrics, chDescs []ChannelDescriptor) *wdrrQueue {
+	// Copy the channel IDs from the channel descriptors and create a WDRR flow
+	// per ChannelDescriptor.
+	chIDs := make([]ChannelID, len(chDescs))
+	flows := make(map[ChannelID]*drrFlow)
+	for i, chDesc := range chDescs {
+		chID := ChannelID(chDesc.ID)
+		chIDs[i] = chID
+
+		capacity := chDesc.MaxSendCapacityBytes
+		if capacity == 0 {
+			capacity = chDesc.MaxSendBytes
+		}
+
+		flows[chID] = &drrFlow{
+			capacity: capacity,
+			quantum:  chDesc.MaxSendBytes,
+			buffer:   make([]Envelope, 0),
+		}
+	}
+
+	return &wdrrQueue{
+		logger:  logger.With("peer", pID),
+		peerID:  pID,
+		metrics: m,
+		chIDs:   chIDs,
+		flows:   flows,
+
+		closer:        tmsync.NewCloser(),
+		enqueueCh:     make(chan Envelope),
+		enqueueCloser: tmsync.NewCloser(),
+		dequeueCh:     make(chan Envelope),
+		dequeueCloser: tmsync.NewCloser(),
+	}
+}
+
+// enqueue returns an unbuffered write-only channel which a producer can send on.
+func (q *wdrrQueue) enqueue() chan<- Envelope {
+	return q.enqueueCh
+}
+
+// dequeue returns an unbuffered read-only channel which a consumer can read from.
+func (q *wdrrQueue) dequeue() <-chan Envelope {
+	return q.dequeueCh
+}
+
+func (q *wdrrQueue) closed() <-chan struct{} {
+	return q.closer.Done()
+}
+
+// close closes the WDRR queue. After this call enqueue() will block, so the
+// caller must select on closed() as well to avoid blocking forever. The
+// enqueue() and dequeue() along with the internal channels will not be closed.
+// Note, close() will block until all externally spawned goroutines have exited.
+func (q *wdrrQueue) close() {
+	q.closer.Close()
+	<-q.enqueueCloser.Done()
+	<-q.dequeueCloser.Done()
+}
+
+// start starts the WDRR queue process. This must be called before the queue can
+// start to process and accept Envelopes.
+func (q *wdrrQueue) start() {
+	go q.enqueueLoop()
+	go q.dequeueLoop()
+}
+
+// enqueueLoop performs a blocking process where a peer's flow, i.e. a p2p Channel,
+// can send an Envelope via the channel returned by enqueue(). This process
+// blocks and continuously reads off of the enqueue channel. For every Envelope
+// read, the corresponding flow is found via the Envelope's ChannelID. If the
+// flow has capacity to accept the Envelope, it is enqueued to the flow's buffer.
+// Otherwise, the Envelope is dropped.
+func (q *wdrrQueue) enqueueLoop() {
+	defer q.enqueueCloser.Close()
+
+	for {
+		select {
+		case e := <-q.enqueueCh:
+			chIDStr := strconv.Itoa(int(e.channelID))
+
+			flow := q.flows[e.channelID]
+			if flow == nil {
+				panic(fmt.Sprintf("queue flow not found for channel %s", chIDStr))
+			}
+
+			msgSize := proto.Size(e.Message)
+			q.metrics.PeerQueueMsgSize.With("ch_id", chIDStr).Set(float64(msgSize))
+
+			enqueued, size := flow.tryEnqueue(e)
```

Comment:
Perhaps we need to instead drop the oldest message in the buffer if it's full instead of dropping the new incoming message?

Reply:
I think we should have a queue-wide buffer size, and if the incoming message has a higher priority than any other buffered messages we should drop lower-priority messages. If there's only messages of the same or higher priorities already queued we should just drop the incoming message.